scalding-core/src/main/scala/com/twitter/scalding/typed/HashJoinable.scala

```diff
@@ -50,11 +52,95 @@ trait HashJoinable[K, +V] extends CoGroupable[K, V] with KeyedPipe[K] {
       val newPipe = new HashJoin(
         RichPipe.assignName(mapside.toPipe(('key, 'value))(fd, mode, tup2Setter)),
         Field.singleOrdered("key")(keyOrdering),
-        mapped.toPipe(('key1, 'value1))(fd, mode, tup2Setter),
+        getMappedPipe(fd, mode),
         Field.singleOrdered("key1")(keyOrdering),
         WrappedJoiner(new HashJoiner(joinFunction, joiner)))
 
       //Construct the new TypedPipe
       TypedPipe.from[(K, R)](newPipe.project('key, 'value), ('key, 'value))(fd, mode, tuple2Converter)
     })
+
+  /**
+   * Returns a Pipe for the mapped (rhs) pipe with checkpointing (forceToDisk) applied if needed.
+   * Currently we skip checkpointing if we're confident that the underlying rhs Pipe is persisted
+   * (e.g. a source / Checkpoint / GroupBy / CoGroup / Every) and we have 0 or more Each operator Fns
+   * that are not doing any real work (e.g. Converter, CleanupIdentityFunction)
+   */
+  private def getMappedPipe(fd: FlowDef, mode: Mode): Pipe = {
+    val mappedPipe = mapped.toPipe(('key1, 'value1))(fd, mode, tup2Setter)
+
+    // if the user has turned off auto force right, we fall back to the old behavior and
+    //just return the mapped pipe
+    if (!getHashJoinAutoForceRight(mode) || isSafeToSkipForceToDisk(mappedPipe, mode)) mappedPipe
+    else mappedPipe.forceToDisk
+  }
+
+  /**
+   * Computes if it is safe to skip a force to disk (only if the user hasn't turned this off using
+   * Config.HashJoinAutoForceRight).
+   * If we know the pipe is persisted,we can safely skip. If the Pipe is an Each operator, we check
+   * if the function it holds can be skipped and we recurse to check its parent pipe.
+   * Recursion handles situations where we have a couple of Each ops in a row.
+   * For example: pipe.forceToDisk.onComplete results in: Each -> Each -> Checkpoint
+   */
+  private def isSafeToSkipForceToDisk(pipe: Pipe, mode: Mode): Boolean = {
+    pipe match {
+      case eachPipe: Each =>
+        if (canSkipEachOperation(eachPipe.getOperation, mode)) {
+          //need to recurse down to see if parent pipe is ok
+          getPreviousPipe(eachPipe).exists(prevPipe => isSafeToSkipForceToDisk(prevPipe, mode))
```

Comment:
Based on my limited understanding of this, it seems as though we shouldn't end up with an Each without an previous Pipe. Can we say that's true (or am I missing something)? 

Regarding the case with two previous pipes - I don't think I've encountered a Pipe with more than one previous. I've seen us handling it in other places in the code but I'm not sure if it was just safe programming (as `previous` is an array and could in theory have multiple elements) or due to a specific scenario. If anyone knows of a situation where this occurs, do chime in. 

I have seen `previous` end up being null though and that is in the case where it is a Splice (the parents there are in the [pipes](https://github.com/twitter/cascading/blob/2.1/cascading-core/src/main/java/cascading/pipe/Splice.java#L62) field, `previous` and `parent` are null. 

Guess, I'm wondering if we do end up forcing to disk here - is that bad? We're not able to infer that the pipe is persisted, so we just end up forcing to disk. We have a config option that allows users to opt out.